state/migration_import.go

```diff
@@ -1407,35 +1407,45 @@ func (i *importer) makeApplicationDoc(a description.Application) (*applicationDo
 	return appDoc, nil
 }
 
+// makeCharmOrigin returns the charm origin for an application
+//
+// Previous versions of the Juju server and clients have treated applications charm
+// origins very loosely, particularly during switches. The server performed no validation
```

Comment:
What is a "switch" in this context? It's `refresh --switch` and not used in every day juju lexicon.